llvm/labs/lab2/safarov_nurlan/replaceMulWithBitShift.cpp

```diff
@@ -0,0 +1,84 @@
+#include "llvm/IR/Function.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include <vector>
+
+using namespace llvm;
+
+namespace {
+
+bool runPlugin(Function &F) {
+  std::vector<Instruction *> instrDeleteList;
+  for (llvm::BasicBlock &funcBlock : F) {
+    for (llvm::Instruction &instr : funcBlock) {
+      if (llvm::BinaryOperator *binOper = dyn_cast<BinaryOperator>(&instr)) {
+        if (binOper->getOpcode() == llvm::Instruction::BinaryOps::Mul) {
+          Constant *constant = dyn_cast<Constant>(binOper->getOperand(0));
+          int notConstantOperandNumber;
+          if (constant) {
+            notConstantOperandNumber = 1;
+          } else {
+            constant = dyn_cast<Constant>(binOper->getOperand(1));
+            notConstantOperandNumber = 0;
+          }
+          if (constant) {
+            llvm::APInt value = constant->getUniqueInteger();
+            if ((value & (value - 1)) == 0) {
+              IRBuilder<> Builder(binOper);
+              Value *newInstr = Builder.CreateShl(
+                  binOper->getOperand(notConstantOperandNumber),
+                  value.logBase2());
+              binOper->replaceAllUsesWith(newInstr);
+              //  errs() << "Replacement: " << instr << "\n";
+              errs() << instr << "\n";
+              instrDeleteList.push_back(&instr);
+            }
+          }
+        }
+      }
+    }
+  }
+  for (auto *instr : instrDeleteList)
+    instr->eraseFromParent();
+
+  return false;
+}
+
+struct RunOrNonePass : PassInfoMixin<RunOrNonePass> {
+  PreservedAnalyses run(Function &F, FunctionAnalysisManager &) {
+    if (!runPlugin(F))
+      return PreservedAnalyses::all();
+    return PreservedAnalyses::none();
```

Comment:
Is your pass so dramatically changes IR that all analysis passes should be re-evaluated?

Reply:
1. **Significant changes in IR:** Pass replaces multiplication operations with bit shifts, which can lead to significant changes in the structure and behavior of the program. This may affect the results of analyses suggesting a certain IR structure.
2. **The need for recalculation:** Since the pass makes such significant changes, LLVM analyses may need to be recalculated to take into account the new IR structure and semantics. Without this recalculation, the analysis results may be inaccurate or incomplete.
3. **Saving analyses:** When returning Preserve Analysis::all(), the pass informs that all analyses must be saved and recalculated. This is important to ensure the correctness and reliability of subsequent optimizations and analyses using IR.
4. **Return PreservedAnalyses::none():** If the pass has not made any changes to the IR, then you can return PreservedAnalyses::none(), which means that the results of all analyses are saved and do not require recalculation.

Reply:
Firstly, I reviewed this code improperly, so in case of changes are made it preserves analyses results. But if something goes wrong (your pass doesn't handle such cases, so else branch is unreachable formally) and pass cannot perform modifications then all analyses should be reset. 
Could you please clarify two sentences, they confuse me a bit:
"When returning Preserve Analysis::all(), the pass informs that all analyses **must be saved and recalculated**." So, should they be saved or should be recalculated?
And what actually `PreservedAnalyses::none()` means?

Reply:
I apologize for any confusion caused by my previous explanation. Let me clarify the points regarding the preservation and recalculation of analyses:

Significant changes in IR: The pass replaces multiplication operations with bit shifts, which can lead to significant changes in the structure and behavior of the program. This affects various analyses, such as data flow analysis and dependence analysis, which are sensitive to the specific instructions present in the IR.

The need for recalculation: Since the pass makes significant changes to the IR, it is crucial to recalculate the analyses to reflect the updated structure and semantics of the IR. Without recalculation, the results of these analyses may become inaccurate or incomplete, leading to potential issues in subsequent optimization passes.

Returning PreservedAnalyses::none(): When the pass modifies the IR, we should return PreservedAnalyses::none(), indicating that all analyses need to be recomputed. This ensures the correctness and reliability of future analyses and optimizations.

Returning PreservedAnalyses::all(): If the pass does not make any changes to the IR, we can return PreservedAnalyses::all(), which means that the results of all analyses are still valid and do not require recalculation.

To address your specific question: When returning PreservedAnalyses::all(), the pass indicates that no changes were made to the IR, so all analyses remain valid and do not need to be recalculated. Conversely, if we return PreservedAnalyses::none(), it means that changes were made, and therefore all analyses should be recomputed to account for the updated IR.

Reply:
Thank you for detailed explanation.